### PR TITLE
All the NoSpace() errors will be handled by regular SetBGError and RecoverFromNoSpace()

### PR DESCRIPTION
**Pull request by zhichao-cao**
In the current logic, any IO Error with retryable flag == true will be handled by the special logic and in most cases, StartRecoverFromRetryableBGIOError will be called to do the auto resume. If the NoSpace error with retryable flag is set during WAL write, it is mapped as a hard error, which will trigger the auto recovery. During the recover process, if write continues and append to the WAL, the write process sees that bg_error is set to HardError and it calls WriteStatusCheck(), which calls SetBGError() with Status (not IOStatus). This will redirect to the regular SetBGError interface, in which recovery_error_ will be set to the corresponding error. With the recovery_error_ set, the auto resume thread created in StartRecoverFromRetryableBGIOError will keep failing as long as user keeps trying to write.

To fix this issue. All the NoSpace error (no matter retryable flag is set or not) will be redirect to the regular SetBGError, and RecoverFromNoSpace() will do the recovery job which calls SstFileManager::StartErrorRecovery().

test plan: make check and added the new testing case

### CODE REVIEW

**Comment by mrambacher** [comment]
I am not sure I understand this change.  What if the EncryptedEnv is used, which wraps a Default one.  Why would it be different?  What about other Env/FileSystems that use/wrap the base ones?

Also, the title says "PosixEnv" but this also applies to WinEnv.  Is it intentional?

Can you also add a comment in the code as to why this change is necessary and what problem it is addressing?

**Comment by anand1976** [comment]
@zhichao-cao The issue is Status::NoSpace() errors from PosixEnv are currently handled in https://github.com/facebook/rocksdb/blob/master/db/error_handler.cc#L380 because it has the retryable flag set in IOStatus, rather than the old SetBGError() https://github.com/facebook/rocksdb/blob/master/db/error_handler.cc#L272, which has some special case handling for out of space errors.

**Comment by zhichao-cao** [comment]
> @zhichao-cao The issue is Status::NoSpace() errors from PosixEnv are currently handled in https://github.com/facebook/rocksdb/blob/master/db/error_handler.cc#L380 because it has the retryable flag set in IOStatus, rather than the old SetBGError() https://github.com/facebook/rocksdb/blob/master/db/error_handler.cc#L272, which has some special case handling for out of space errors.

I see. Will change accordingly.

**Comment by zhichao-cao** [comment]
> I am not sure I understand this change. What if the EncryptedEnv is used, which wraps a Default one. Why would it be different? What about other Env/FileSystems that use/wrap the base ones?
> 
> Also, the title says "PosixEnv" but this also applies to WinEnv. Is it intentional?
> 
> Can you also add a comment in the code as to why this change is necessary and what problem it is addressing?

The PR is updated, the previous change was a wrong fix remove all the title, descriptions, and the code.

File: db/error_handler.cc [hunk 1]
```diff
@@ -418,15 +418,18 @@ const Status& ErrorHandler::SetBGError(const IOStatus& bg_io_err,
                                           &bg_err, db_mutex_, &auto_recovery);
     recover_context_ = context;
     return bg_error_;
-  } else if (bg_io_err.GetScope() ==
-                 IOStatus::IOErrorScope::kIOErrorScopeFile ||
-             bg_io_err.GetRetryable()) {
+  } else if (bg_io_err.subcode() != IOStatus::SubCode::kNoSpace &&
```
Comment (anand1976) [inline db/error_handler.cc:421]: Interesting. Just realized that the ```==``` operator of ```Status``` and ```IOStatus``` only checks code and not subcode. So is there a potential bug in line 330?

**Review by anand1976 — COMMENTED** [review]
Thanks for the fix

**Review by anand1976 — APPROVED** [review]
The change in this PR looks good. The potential bug I mentioned in the comment needs to be looked at, but its most likely benign since NoSpace() is the only error that's marked as a soft error and retried in the old ```SetBGError```.

File: db/error_handler.cc (hunk 1 above)
Comment (zhichao-cao) [reply db/error_handler.cc:421]: I will check line 330. I just realized the == operator only compares the code not subcode when did the implementation

**Comment by zhichao-cao** [comment]
> The change in this PR looks good. The potential bug I mentioned in the comment needs to be looked at, but its most likely benign since NoSpace() is the only error that's marked as a soft error and retried in the old `SetBGError`.

I will create another PR to fix it. Some of the testing cases needs to be rewrite.

**Comment by ajkr** [comment]
It's an interesting fix. Just a general question - why can foreground writes set recovery_error_? Can other hard/retryable/file-scoped errors run into the same problem?

**Pull request closed without merging**